include/deal.II/multigrid/mg_transfer_global_coarsening.templates.h

```diff
@@ -865,28 +878,46 @@ namespace internal
       IndexSet is_dst_remote(this->cell_id_translator.size());
       IndexSet is_src_locally_owned(this->cell_id_translator.size());
 
-      for (const auto &cell : tria_dst.active_cell_iterators())
-        if (cell->is_locally_owned())
-          is_dst_locally_owned.add_index(
-            this->cell_id_translator.translate(cell));
+      const auto fine_operation = [&](const auto &cell) {
+        is_dst_locally_owned.add_index(
+          this->cell_id_translator.translate(cell));
+      };
 
+      const auto coarse_operation = [&](const auto &cell) {
+        is_src_locally_owned.add_index(
+          this->cell_id_translator.translate(cell));
 
-      for (const auto &cell : tria_src.active_cell_iterators())
-        if (cell->is_locally_owned())
-          {
-            is_src_locally_owned.add_index(
-              this->cell_id_translator.translate(cell));
-            is_dst_remote.add_index(this->cell_id_translator.translate(cell));
+        // in the case of global coarsening identity transfer is possible
+        if (mg_level_coarse == numbers::invalid_unsigned_int)
+          is_dst_remote.add_index(this->cell_id_translator.translate(cell));
 
-            if (cell->level() + 1u == tria_dst.n_global_levels())
-              continue;
+        if (cell->level() + 1u == tria_dst.n_global_levels())
+          return;
 
-            for (unsigned int i = 0;
-                 i < GeometryInfo<dim>::max_children_per_cell;
-                 ++i)
-              is_dst_remote.add_index(
-                this->cell_id_translator.translate(cell, i));
+        for (unsigned int i = 0; i < GeometryInfo<dim>::max_children_per_cell;
+             ++i)
+          is_dst_remote.add_index(this->cell_id_translator.translate(cell, i));
+      };
+
+      const auto loop = [&](const auto &tria,
+                            const auto  level,
+                            const auto &op) {
+        if (level == numbers::invalid_unsigned_int)
+          {
+            for (const auto &cell : tria.active_cell_iterators())
+              if (!cell->is_artificial() && cell->is_locally_owned())
+                op(cell);
           }
+        else
+          {
+            for (const auto &cell : tria.cell_iterators_on_level(level))
+              if (cell->level_subdomain_id() == tria.locally_owned_subdomain())
+                op(cell);
+          }
+      };
```

Comment:
I do not really like the nested use of several lambda functions here as it makes it somewhat complicated to see what is going on, but I admit that I don't see a simple solution without code duplication or collection of the cells to loop over in some different way, so let's keep it like this.